crates/rpc/src/jsonrpc/router.rs

```diff
@@ -885,4 +953,93 @@ mod tests {
 
         assert_eq!(content_type, "application/json");
     }
+
+    #[tokio::test]
+    async fn concurrent_futures() {
```

Comment:
How reliable do you think this test is since its relying on time?

There is probably something more complicated one could do to force the order of execution. For example, by linking the futures together using channels or some form of reference counting.

e.g. each future has a sender and receiver which in turn connects to another future. The final future's receiver is connected to the outside sender. The future does `_ = rx.recv().await`. From the outside we then drop the sender, which causes the future to complete, in turn dropping the sender the future is holding. This in turn causes the next future to complete etc.

Reply:
+1

Reply:
Thanks the the feedback and suggestion. I rewrote this using channels & [Tokio's Notify](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html).